ouroboros-network-protocols/testlib/Ouroboros/Network/Protocol/KeepAlive/Direct.hs

```diff
@@ -1,19 +1,28 @@
 {-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RankNTypes     #-}
 
 module Ouroboros.Network.Protocol.KeepAlive.Direct where
 
 import           Ouroboros.Network.Protocol.KeepAlive.Client
 import           Ouroboros.Network.Protocol.KeepAlive.Server
 
-direct :: Monad m
+
+direct :: forall a b m. Monad m
        => KeepAliveServer m a
        -> KeepAliveClient m b
        -> m (a, b)
-direct KeepAliveServer { recvMsgDone }
-       (SendMsgDone mdone) =
-    (,) <$> recvMsgDone <*> mdone
-direct KeepAliveServer { recvMsgKeepAlive }
-       (SendMsgKeepAlive _cookie mclient) = do
-    server <- recvMsgKeepAlive
-    client <- mclient
-    direct server client
+direct srv (KeepAliveClient clientM) = do
+  go srv =<< clientM
+ where
+  go :: Monad m
+     => KeepAliveServer m a
+     -> KeepAliveClientSt m b
+     -> m (a, b)
+  go KeepAliveServer { recvMsgDone }
+         (SendMsgDone mdone) =
+      (,) <$> recvMsgDone <*> mdone
+  go KeepAliveServer { recvMsgKeepAlive }
+         (SendMsgKeepAlive _cookie mclient) = do
+      server <- recvMsgKeepAlive
+      client <- mclient
+      go server client
```

Comment:
indentation & use of applicative instance rather than `Monad`.
```suggestion
direct srv (KeepAliveClient clientM) = do
   go srv =<< clientM
 where
   go :: Monad m
      => KeepAliveServer m a
      -> KeepAliveClientSt m b
      -> m (a, b)
   go KeepAliveServer { recvMsgDone }
          (SendMsgDone mdone) =
       (,) <$> recvMsgDone <*> mdone
   go KeepAliveServer { recvMsgKeepAlive }
          (SendMsgKeepAlive _cookie mclient) =
       go <$> recvMsgKeepAlive
          <*> mclient
```